File: Modulo10/UniversidadeDepartamento-Java/Universidade.java

public class Universidade {

    private String nome;
    private Departamento departamentos[];
    private int qtde;
    private int max;

    public Universidade(String nome) {
        System.out.println("[Construindo " + nome + "]");
        this.nome = nome;
        this.qtde = 0;
        this.max = 2;
        departamentos = new Departamento[max];
    }

    public String getNome() {
        return this.nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    private void realoca(int tam) {

        // cria um novo array com mudança no tamanho do array

        max = tam;

        Departamento aux[] = new Departamento[max];

        for (int i = 0; i < qtde; i++) {
            aux[i] = departamentos[i];
        }

        departamentos = aux;
    }

    public void adicionaDepartamento(String nome, String sigla) {
        if (qtde == max) {
            realoca(max + 5);
        }
        departamentos[qtde++] = new Departamento(nome, sigla, this);
    }

    public void removeDepartamento(String sigla) {
        boolean found = false;
        int i = 0;

        while (i < qtde && !found) {
            if (departamentos[i].getSigla().equals(sigla)) {
                found = true;
            } else {
                i++;
            }
        }

        if (found) {
            departamentos[i] = null;

            while (i < qtde) {
                departamentos[i] = departamentos[i + 1];
                i++;
            }

            qtde = qtde - 1;

            if (qtde == max - 5) {
                realoca(max - 5);
            }
        }
    }

    public void imprime() {
        System.out.println("Departamentos da " + nome);
        for (int i = 0; i < qtde; i++) {
            departamentos[i].imprime();
        }
    }
}

File: Modulo07/UniversidadeDepartamento/Universidade.cpp
#include "Universidade.h"

Universidade::Universidade(string nome) :
nome(nome) {
    cout << "[Construindo " << nome << "]" << endl;
    qtde = 0;
    max = 2;
    departamentos = new Departamento*[max];
}

Universidade::~Universidade() {
    for (int i = 0; i < qtde; i++) {
        delete departamentos[i];
    }
    delete departamentos;
    cout << "[Destruindo " << nome << "]" << endl;
}

string Universidade::getNome() const {
    return nome;
}

void Universidade::setNome(string nome) {
    this->nome = nome;
}

void Universidade::realoca(int tam) {

    // cria um novo array com mudança no tamanho do array

    max = tam;

    Departamento** aux = new Departamento*[max];

    for (int i = 0; i < qtde; i++) {
        aux[i] = departamentos[i];
    }

    delete departamentos; // liberando o antigo array
    departamentos = aux;
}

void Universidade::adicionaDepartamento(string nome, string sigla) {
    if (qtde == max) {
        realoca(max + 5);
    }
    departamentos[qtde++] = new Departamento(nome, sigla, this);
}

void Universidade::removeDepartamento(string sigla) {
    bool found = false;
    int i = 0;

    while (i < qtde && !found) {
        // if (departamentos[i]->getSigla().compare(sigla) == 0) {
        if (departamentos[i]->getSigla() == sigla) {
            found = true;
        } else {
            i++;
        }
    }

    if (found) {
        delete departamentos[i];

        while (i < qtde) {
            departamentos[i] = departamentos[i + 1];
            i++;
        }

        qtde = qtde - 1;

        if (qtde == max - 5) {
            realoca(max - 5);
        }
    }
}

void Universidade::imprime() const {
    cout << "Departamentos da " << nome << endl;
    for (int i = 0; i < qtde; i++) {
        departamentos[i]->imprime();
    }
}